src/inferenceql/query/cache.cljc

```diff
@@ -0,0 +1,20 @@
+(ns inferenceql.query.cache
+  "For caching expensive results."
+  (:require #?(:clj [clojure.core.memoize :as memo]
+               :cljs [memoizee :as memoizee])
+            [clojure.string :as string]))
+
+(def default-threshold 100)
+
+
+(defn lru
+  "Memoizes a fn with a least-recently-used eviction policy.
+
+  After the number of cached results exceeds the threshold, the
+  least-recently-used ones will be evicted."
+  ([f]
+   (lru f default-threshold))
+  ([f lru-threshold]
+   #?(:clj (memo/lru f :lru/threshold lru-threshold)
+      :cljs (memoizee f #js {"max" lru-threshold
+                             "normalizer" js/JSON.stringify}))))
```

Comment:
I wonder if we'll run into trouble using `js/JSON.stringify` with some of the parameters passed to these functions, like `model` (which can be a special type, eg. [reify](https://github.com/InferenceQL/inferenceql.query/blob/9cd372e042780cac173e57d2c46eb7df1ac02698/test/inferenceql/query/scalar_test.cljc#L253)) or cljs data types which look pretty weird (though maybe that's ok, if it's more performant than using `clj->js` first..)

<img width="1178" alt="image" src="https://github.com/InferenceQL/inferenceql.query/assets/165223/3f5bdec3-07eb-4530-802b-4f6a433e25a8">

It might be worth doing some simple benchmarks before committing to an approach, as some of these conversions can be surprisingly costly.

Reply:
My predictions are that it should be fine, but it won't hurt to check. Will have to wait until I get back from vacation, though.

The serialized weirdness should be ok as long as they never result in accidentally identical string representations. If anything, I think the opposite is true. There are probably things that could share keys, but won't. Luckily, that just means some cache misses.

I would be _very_ surprised if clj->js + JSON/stringify was faster than JSON/stringify, but no need to guess! I'll try it out when I get back. We'll see what my posteriors are then.

Reply:
Well, it turns out I am very surprised. Unfortunately, I don't think it will work out to use clj->js first.

I ran a quick-and-dirty test in cljs:
```clojure
(let [n 100000
        m {"x" 0 :foo {:bar 1 :moop/floop "asdf"} :baz [1 2 3]}
        a (reify IAtom)]

    (js/console.log "JSON.stringify:")
    (time
      (dotimes [_ n]
        (js/JSON.stringify m)
        (js/JSON.stringify :foo)
        (js/JSON.stringify a)))

    (js/console.log "clj->js, then JSON.stringify:")
    (time
      (dotimes [_ n]
        (-> m clj->js (js/JSON.stringify))
        (-> :foo clj->js (js/JSON.stringify))
        (-> a clj->js (js/JSON.stringify))))

    (js/console.log "clj->js w/ str keyword-fn, then JSON.stringify:")
    (time
      (dotimes [_ n]
        (-> m (clj->js :keyword-fn str) (js/JSON.stringify))
        (-> :foo (clj->js :keyword-fn str) (js/JSON.stringify))
        (-> a (clj->js :keyword-fn str) (js/JSON.stringify))))

    (js/console.log "bean/->js, then JSON.stringify:")
    (time
      (dotimes [_ n]
        (-> m (bean/->js :key->prop str) (js/JSON.stringify))
        (-> :foo (bean/->js :key->prop str)(js/JSON.stringify))
        (-> a (bean/->js :key->prop str) (js/JSON.stringify)))))
```
and got:
```
JSON.stringify:
"Elapsed time: 1274.645467 msecs"
clj->js, then JSON.stringify:
"Elapsed time: 1208.113983 msecs"
clj->js w/ str keyword-fn, then JSON.stringify:
"Elapsed time: 1443.455173 msecs"
bean/->js, then JSON.stringify:
"Elapsed time: 1506.302295 msecs"
```

Using clj->js before JSON.stringify _is_ slightly faster, but unfortunately, it can't distinguish between string and keyword keys. Having duplicate string and keyword keys would cause other problems, so it might be safe to do. But, I'd rather err on the side of correctness, especially since these timings aren't too far off from each other.

We can always revisit our caching strategy, if necessary.